[MRG] Relaxed criterion specification so tree plot does not force calculate entropy for any DecisionTreeClassifier

fhoang7 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#pull-request-checklist
-->

#### Reference Issues/PRs
Fixes #13944 


#### What does this implement/fix? Explain your changes.
Removed hard coding of criterion in tree making function. Instead, passed in criterion as an additional input into the function and modified subsequent function calls accordingly. export.py test file was modified to ensure that all tests passed as a result of this change. 

#### Any other comments?
Need feedback on my approach and potential next steps. I would assume that the next thing to do would be to write a separate test to ensure that the proper decision tree criteria will be printed when entropy is the desired metric. 

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

Commit:
Relaxed criterion specification so plot_tree no longer writes entropy for any criterion in decision tree

Commit:

Please enter the commit message for your changes. Lines starting

Commit:
Revert " Please enter the commit message for your changes. Lines starting"

This reverts commit 9a7ba92d

Commit:
Ensure changes conform to flake8 character limits

Commit:
Resolved last line that violated pep8

Commit:
Add test to test that entropy is calculated & displayed in tree diagram when decision tree classifier computes entropy

abhinavsagar commented on sklearn/tree/tests/test_export.py:
```diff
@@ -399,20 +399,39 @@ def test_export_text():
     nodes = plot_tree(clf, feature_names=feature_names)
     assert len(nodes) == 3
-    assert nodes[0].get_text() == ("first feat <= 0.0\nentropy = 0.5\n"
+    assert nodes[0].get_text() == ("first feat <= 0.0\nentropy = 1.0\n"
```
Could you remove entropy and instead use criterion. Else LGTM.

fhoang7 replied:
Will make that change. Thanks!

jnothman replied:
@abhinavsagar this test is for the case where the tree's criterion is indeed entropy. See test_plot_tree_gini below.

jnothman reviewed (approved):
Thanks!

I think we'd like to release this in 0.21.3, but it's not clear to me whether that release will be necessary, or whether we should just put this in 0.22 for now. I'd say, for now, we'll create the 0.21.3 section, because it's still very early to exclude needing another patch release.

Please merge master and add an entry to the 0.21.3 change log at `doc/whats_new/v0.21.rst`. Like the other entries there, please reference this pull request with `:pr:` and credit yourself (and other contributors if applicable) with `:user:`.

Commit:
Update v0.21.rst

Added #13947 to change log

fhoang7 commented:
@jnothman How do I merge master? I thought that someone with write access could do it.

jnothman commented:
I meant to fetch the latest version of master, merge it into your branch,
commit and push. There are some instructions at
https://scikit-learn.org/dev/developers/contributing.html

Commit:
Merge remote-tracking branch 'upstream/master' into my-feature

fhoang7 commented:
Merged master and added to change log like you suggested. Is there anything else I need to follow up @jnothman or is it out of my hands now? Also, wasn't sure if I added to the change log correctly. Didn't see anybody else make pull requests to amend the change log.

jnothman commented:
I was hoping you would amend the change log in this pull request (you can just merge your changs log branch into this one). The reason for merging master was only that I'd changed the change log.

Commit:
Merge pull request #1 from fhoang7/patch-1

Update v0.21.rst

fhoang7 commented:
Done! Merged my other request into this one.

Commit:
Update splitting.pyx

Initialized gain variable with parent node negative loss and removing from _split_gain function in order to improve performance of splitting.

Commit:
Fix global variable accessing

Commit:
Revert "Fix global variable accessing"

This reverts commit 7011bb5b00b46d8e39f380b044f95c9ede293189.

Commit:
Revert "Update splitting.pyx"

This reverts commit d66c9d4f7ffeb3c4314d722d4c4a6e32779edcd5.

fhoang7 commented:
Reverted 2 commits since I was developing for another issue on this branch. Moved those changes to a separate branch.

Commit:
Move change log update into proper spot and reference public API

fhoang7 commented on doc/whats_new/v0.21.rst:
```diff
@@ -885,6 +885,10 @@ Support for Python 3.4 and below has been officially dropped.
   ``max_depth`` were both specified by the user. Please note that this also
   affects all ensemble methods using decision trees.
   :pr:`12344` by :user:`Adrin Jalali <adrinjalali>`.
+  
```
Addressed in commit

fhoang7 commented on doc/whats_new/v0.21.rst:
```diff
@@ -885,6 +885,10 @@ Support for Python 3.4 and below has been officially dropped.
   affects all ensemble methods using decision trees.
   :pr:`12344` by :user:`Adrin Jalali <adrinjalali>`.
+  
+- |Fix| Fixed an issue with :class:`MLPTreeExporter` where it display
```
Addressed in commit

jnothman reviewed (commented):
Thanks @fhoang7

amueller commented:
Thanks!
